feat: add Terrain Variation

davo0411 opened the pull request:
This is the final version of my plugin, Terrain Variation.
It fixes the in game texture tiling visible on terrain via Stochastic Texturing.
Works with vanilla, PBR, Complex Materials and Parallax.
Untested with CS Terrain Helper.
Untested in VR, no clue if that has a different pipeline for terrain or anything.

How it works:
1. Transforms UV coordinates into a triangular grid space using a skew transformation matrix, creating a more natural-looking pattern than square grid sampling.
2. Uses barycentric coordinates (the barry variable) to interpolate between three chosen points in this grid.
3. Applies hash-based randomisation with the hash2D2D function to create pseudo-random but deterministic offsets for each sample point.
Some blending is applied, and the offsets are applied to all texture layers the same, resulting in the texture being offset by the correct, same amount, for each layer.

![BeforeVolcanic](https://github.com/user-attachments/assets/ac86ef56-45f6-49a5-998a-3afa43df6330)
![AfterVolcanic](https://github.com/user-attachments/assets/59b33103-28ef-4f0c-914a-c8bdc18b3070)

![BeforeTundra](https://github.com/user-attachments/assets/eae310d6-c6b8-43f3-a375-957300690c5f)
![AfterTundra](https://github.com/user-attachments/assets/471f9c81-55d3-4f9f-83c7-6233cb06e9ea)

![Performance](https://github.com/user-attachments/assets/a5de31b0-0cb5-4acb-a707-363650348601)
![Performance 2](https://github.com/user-attachments/assets/cc47a4b9-960f-41ba-ad0e-e715750bf2f5)

davo0411 committed:
Update Lighting.hlsl

davo0411 committed:
Added a Terrain Tiling Fix to Lighting.hlsl

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Update lighting.hlsl

Fixed small compile issue with PBR. No difference to visual effect.

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
First of several fixes

davo0411 committed:
Merge branch 'terrain-variation' of https://github.com/davo0411/skyrim-community-shaders into terrain-variation

davo0411 committed:
Update Lighting.hlsl

davo0411 committed:
Separate Feature Beta 1

Working Terrain Tiling Fix as a separate feature within Community Shaders. Has a on/off switch in the settings.

davo0411 committed:
Updated Split plugin

Works with or without terrainvariation plugin enabled now

davo0411 committed:
Update

davo0411 committed:
Fix for Parallax Shadows on Whole Terrain

Fix for weird issue with parallax terrain shadows causing the whole terrain to go black at certain distances.

davo0411 committed:
FINAL

Plugin now works fully when installed, and does not throw any errors when TerrainVariation.hlsli is not present. This allows the plugin to be a separate nexus release.

davo0411 committed:
Comment Cleanup 1

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Delete .gitignore

davo0411 committed:
Update ExtendedMaterials.hlsli

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 commented on package/Shaders/Lighting.hlsl:
```diff
@@ -1233,18 +1238,25 @@ PS_OUTPUT main(PS_INPUT input, bool frontFace
-	input.LandBlendWeights1.y /= landBlendWeights;
-	input.LandBlendWeights1.z /= landBlendWeights;
-	input.LandBlendWeights1.w /= landBlendWeights;
-	input.LandBlendWeights1.x /= landBlendWeights;
```
double division of 1.x that didn't really make sense to me and sorta made the effect weird in my very early testing. Might be an unnecessary change, but i think the new works better.

davo0411 commented on package/Shaders/Lighting.hlsl:
```diff
@@ -1300,49 +1321,286 @@ PS_OUTPUT main(PS_INPUT input, bool frontFace
+#		if defined(TRUE_PBR)
+	rawRMAOS = TexRMAOSSampler.SampleBias(SampRMAOSSampler, diffuseUv, SharedData::MipBias) * float4(PBRParams1.x, 1, 1, PBRParams1.z);
+#		endif
+#	endif
```
essentially streamlining the terrain layers, the slightly tweaked order didn't seem to break anything but makes the code a lot easier to read through.

davo0411 commented on package/Shaders/Lighting.hlsl:
```diff
@@ -2225,30 +2375,30 @@ PS_OUTPUT main(PS_INPUT input, bool frontFace
+		// 				parallaxShadow = ExtendedMaterials::GetParallaxSoftShadowMultiplier(uv, mipLevel, lightDirectionTS, sh0, TexParallaxSampler, SampParallaxSampler, 0, parallaxShadowQuality, screenNoise, displacementParams);
+		// #				endif
+		// 		}
+		// #			endif
```
this section i believe is now being calculated above. Probably needs further testing, but threw errors and broke it last i had it enabled. have had commented out since early testing, so might be unnecessary.

davo0411 reviewed (commented):
some annotations to my lighting.hlsl changes

davo0411 committed:
fixing gitignore

alandtse commented on package/Shaders/Common/StochasticSampling.hlsli:
```diff
@@ -0,0 +1,74 @@
+// Stochastic sampling implementation for texture tiling reduction
+// Based on papers "Rendering Antialiased Edges with Barycentric Textures" and "Tiled Stochastic Texture" by Thomas Deliot & Eric Heitz.
```
Do you have links?

davo0411 replied:
Fixed

alandtse commented on package/Shaders/Lighting.hlsl:
```diff
@@ -2225,30 +2375,30 @@ PS_OUTPUT main(PS_INPUT input, bool frontFace
-#				endif
-		}
-#			endif
+		// #			if defined(EMAT)
```
Do not include dead code. It should be removed if you don't need it.

davo0411 replied:
Fixed. This was old lighting.hlsl code, left it in case it served a purpose beyond what i have essentially replaced/rewritten

alandtse reviewed (changes requested):
In addition to the line comments:

1. Ensure there are no hlsl warnings or errors caused by your changes. You may have to turn on debug mode. To the extent some warnings are there from older code in the files you've touched, consider cleaning them up in a prereq PR.
2. Test with the feature off both in config (if an option) and with disable with boot under advanced.
3. You may post a test build after addressing both 1 and 2. All major errors should be handled before we merge to dev.

davo0411 committed:
Updating References

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 commented:
> In addition to the line comments:
> 
> 1. Ensure there are no hlsl warnings or errors caused by your changes. You may have to turn on debug mode. To the extent some warnings are there from older code in the files you've touched, consider cleaning them up in a prereq PR.
> 2. Test with the feature off both in config (if an option) and with disable with boot under advanced.
> 3. You may post a test build after addressing both 1 and 2. All major errors should be handled before we merge to dev.

1. No errors with my compile, I have tested 4 different states:
AIO compiles fine
AIO With plugin disabled at boot
Individual plugins, with terrain variation disabled in MO2
Individual plugins, with terrain variation enabled in MO2.
1a. I'm not sure what you're saying, are you saying that previous warnings unrelated to my code need to be tidied before i can get my plugin accepted? I don't really get why. There are pre-existing warnings within lighting.hlsl, and i'm not gonna spend a day trying to troubleshoot those when they are unrelated to my changes.

2. Compiles just file
3. I have a test build AIO posted in the discord, davo's Terrain Variation thread in the community channel.

alandtse reviewed (approved):
@doodlum ready for your review pre merge.

alandtse commented:
Please check the names of things. The feature is Terrain Variations but it shows up in the menu as something else.

davo0411 committed:
Name Fix, General Tidy-up

davo0411 commented:
> Please check the names of things. The feature is Terrain Variations but it shows up in the menu as something else.

Fixed this up, sorry was flipping between names but i think the tooltip is enough to explain what it does, and allows me to add more features or "Variation" ideas in future ( i was thinking something with texture-bombing, like puddles or other things onto terrain randomly)

ThePagi commented on package/Shaders/Lighting.hlsl:
```diff
@@ -1300,49 +1321,286 @@ PS_OUTPUT main(PS_INPUT input, bool frontFace
-			baseColor = float4(rawBaseColor.rgb / Color::PBRLightingScale, rawBaseColor.a);
-		}
-		else
+	[branch] if ((PBRFlags & PBR::TerrainFlags::LandTile0PBR) == 0)
```
Why is this still here? It's only relevant to terrain, shouldn't be in non-landscape code.

davo0411 replied:
bcs it break everythang i fix this later loll

alandtse replied:
"Fix this later" means we won't merge unless @ThePagi says to let it through.

davo0411 replied:
I have reworked this slightly, and I need to know from @ThePagi if it works with his snowcover, as that is what it is related to. I haven't messed with the snowcover shader. This segment being below is purely cosmetic, and in brief testing when moving it earlier it broke glint. I can spend the time trying to merge the two areas if absolutely needed.

ThePagi commented on package/Shaders/Lighting.hlsl:
```diff
@@ -1353,255 +1611,138 @@ PS_OUTPUT main(PS_INPUT input, bool frontFace
-	if (input.LandBlendWeights1.y > 0.0) {
-		float4 landColor2 = TexLandColor2Sampler.SampleBias(SampLandColor2Sampler, uv, SharedData::MipBias);
+	// Layer 1 (LandBlendWeights1.x)
+	if (input.LandBlendWeights1.x > 0.0) {
```
Why not put all these together with the diffuse and normal sampling for terrain? Also since you stopped checking if the weight > 0 there, you might be sampling some nonsense texture, which may or may not break something, but no point sampling all that when weight is 0.

davo0411 replied:
fixed :)

davo0411 committed:
Optimisations 1

Added checks for land weights, ensuring that sampling doesn't occur if it isn't needed. Should reduce overhead. Also tidied up the code, incorporating the blending of diffuse/rmaos/alpha into the layer code.

davo0411 committed:
Optimisations 1.1

davo0411 committed:
style: 🎨 apply clang-format changes

doodlum commented:
Prior to merge I need to see a performance comparison, e.g. https://www.capframex.com/

davo0411 committed:
Attempt 1 to support terrain helper

davo0411 committed:
Merge branch 'dev' into Terrain-Variation-ASFEATURE

davo0411 committed:
Merge branch 'dev' into Terrain-Variation-ASFEATURE

davo0411 committed:
Parallax de-borking

Basically reverting a lot of stuff i broke while testing back to original code

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Tidy up

davo0411 committed:
style: 🎨 apply clang-format changes

alandtse commented:
Also, please see if using the new functions from [this commit](https://github.com/doodlum/skyrim-community-shaders/commit/23c9c468c9cd1778fa8badc86f7846c4c77f6bcf) are useful for you.

davo0411 commented:
> Also, please see if using the new functions from [this commit](https://github.com/doodlum/skyrim-community-shaders/commit/23c9c468c9cd1778fa8badc86f7846c4c77f6bcf) are useful for you.

That is very cool, I'll incorporate that in time.
I have done a very large rewrite of the code on suggestion of exist and a couple others who were testing. The function now works on a distance-based slider, a video in action is visible in discord. Idk if its best to do a new PR or merge it into this (i created a separate branch as i didn't know if it was gonna work)
